File: src/main/java/ro/nicolaemariusghergu/queryit/model/Product.java
package ro.nicolaemariusghergu.queryit.model;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Objects;

@Getter
@Setter
@Entity
@ToString
@Table(name = "products")
public class Product implements Serializable {

    @Id
    @Column(name = "product_id", nullable = false)
    private Long id;

    @Column(name = "name")
    private String name;

    @Column(name = "price")
    private Double price;

    @Column(name = "quantity")
    private Integer quantity;

    @ToString.Exclude
    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ToString.Exclude
    @ManyToOne
    @JoinColumn(name = "promotion_id", nullable = false)
    private Promotion promotion;

    @ToString.Exclude
    @ManyToOne
    @JoinColumn(name = "minimarket_id")
    private MiniMarket miniMarket;

    @ToString.Exclude
    @ManyToOne
    @JoinColumn(name = "manufacturer_id", nullable = false)
    private Manufacturer manufacturer;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Product product = (Product) o;
        return id.equals(product.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/main/kotlin/ro/nicolaemariusghergu/queryit/model/Product.kt
package ro.nicolaemariusghergu.queryit.model

import javax.persistence.*
import org.hibernate.annotations.NotFound
import org.hibernate.annotations.NotFoundAction
import java.math.BigDecimal

@Entity
@Table(name = "products")
class Product(
        @Column(name = "name")
        val name: String,

        @Column(name = "price")
        val price: BigDecimal,

        @Column(name = "quantity")
        var quantity: Int,

        @Column(name = "icon_path")
        val iconPath: String,

        @ManyToOne
        @JoinColumn(name = "category_id")
        @NotFound(action = NotFoundAction.IGNORE)
        val category: Category,

        @ManyToOne(cascade = [CascadeType.MERGE])
        @NotFound(action = NotFoundAction.IGNORE)
        @JoinColumn(name = "promotion_id")
        val promotion: Promotion? = null,

        @ManyToOne
        @JoinColumn(name = "minimarket_id")
        @NotFound(action = NotFoundAction.IGNORE)
        val miniMarket: MiniMarket,

        @ManyToOne
        @JoinColumn(name = "manufacturer_id")
        @NotFound(action = NotFoundAction.IGNORE)
        val manufacturer: Manufacturer,

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        @Column(name = "product_id", nullable = false)
        @NotFound(action = NotFoundAction.IGNORE)
        val id: Long? = null
)